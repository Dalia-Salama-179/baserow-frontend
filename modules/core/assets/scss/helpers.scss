.clearfix::after {
  clear: both;
  content: '';
  display: table;
}

.hidden {
  display: none;
}

.visibility-hidden {
  visibility: hidden;
  transform: translateX(-10000px) translateY(-10000px);
}

.align-right {
  text-align: right;
}

.background-white {
  background-color: $white;
}

.color-primary {
  color: $color-primary-500 !important;
}

.color-success {
  color: $color-success-500 !important;
}

.color-warning {
  color: $color-warning-500 !important;
}

.color-error {
  color: $color-error-500 !important;
}

.margin-top-0 {
  margin-top: 0 !important;
}

.margin-top-1 {
  margin-top: 8px !important;
}

.margin-top-2 {
  margin-top: 16px !important;
}

.margin-top-3 {
  margin-top: 30px !important;
}

.margin-top-4 {
  margin-top: 40px !important;
}

.margin-bottom-0 {
  margin-bottom: 0 !important;
}

.margin-bottom-1 {
  margin-bottom: 8px !important;
}

.margin-bottom-2 {
  margin-bottom: 16px !important;
}

.margin-bottom-3 {
  margin-bottom: 30px !important;
}

.margin-bottom-4 {
  margin-bottom: 40px !important;
}

.margin-left-auto {
  margin-left: auto !important;
}

.margin-left-0 {
  margin-left: 0 !important;
}

.margin-left-1 {
  margin-left: 8px !important;
}

.margin-left-2 {
  margin-left: 16px !important;
}

.margin-right-auto {
  margin-right: auto !important;
}

.margin-right-0 {
  margin-right: 0 !important;
}

.margin-right-1 {
  margin-right: 8px !important;
}

.margin-right-2 {
  margin-right: 16px !important;
}

.resizing-horizontal {
  cursor: col-resize;
}

// This selector can only be used to temporarily show an element for calculating
// positions.
.forced-block {
  display: block !important;
}

// Sometimes we forcefully want to disable the the user-select property to initial
// because the element is for example being edited via a contenteditable attribute
// which does not work in combination with the user-select: none; property in Safari.
.forced-user-select-initial {
  user-select: initial !important;
}

.forced-text-overflow-initial {
  text-overflow: initial !important;
}

.prevent-scroll {
  overflow: hidden !important;
}

@keyframes spin {
  0% {
    transform: rotate(0);
  }

  50% {
    transform: rotate(360deg);
  }

  100% {
    transform: rotate(720deg);
  }
}
