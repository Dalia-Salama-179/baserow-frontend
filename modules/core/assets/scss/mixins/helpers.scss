@mixin fixed-height($height, $font-size) {
  height: $height;
  line-height: $height;
  font-size: $font-size;
}

@mixin center-text($width, $font-size, $height: $width) {
  @include fixed-height($height, $font-size);

  width: $width;
  text-align: center;
}

@mixin loading($size: 1.4rem) {
  position: relative;
  display: block;
  width: $size;
  height: $size;
  border-radius: 50%;
  border: 0.25em solid;
  border-color: $color-primary-500 transparent $color-primary-500 transparent;
  animation: spin infinite 1800ms;
  animation-timing-function: cubic-bezier(0.785, 0.135, 0.15, 0.86);
}

// The basically works the same as padding: 10px 0 0 10px or padding: 10px; only then
// resulting position: absolute; with corresponding top, right, bottom, left values.
// -------
// Example:
//  @include absolute(10px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 10px; bottom: 10px; left: 10px;
// -------
// Example:
//  @include absolute(10px, 20px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 10px; left: 20px;
// -------
// Example:
//  @include absolute(10px, 20px, 30px);
// Would result in:
//  position: absolute;
//  top: 10px; right: 20px; bottom: 30px;
@mixin absolute($top: null, $right: null, $bottom: null, $left: null) {
  position: absolute;

  @if ($top != null and $right == null and $bottom == null and $left == null) {
    top: $top;
    right: $top;
    bottom: $top;
    left: $top;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else if ($top != null and $right != null and $bottom == null and
  $left == null) {
    top: $top;
    right: $right;
    bottom: $top;
    left: $right;
  }

  /* stylelint-disable-next-line at-rule-no-unknown */
  @else {
    @if $top != null {
      top: $top;
    }

    @if $right != null {
      right: $right;
    }

    @if $bottom != null {
      bottom: $bottom;
    }

    @if $left != null {
      left: $left;
    }
  }
}
