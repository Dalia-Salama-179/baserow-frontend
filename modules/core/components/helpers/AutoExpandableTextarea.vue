<template>
  <textarea
    ref="inputTextArea"
    :value="value"
    :placeholder="placeholder"
    class="input auto-expandable-textarea"
    :style="{
      height: textBoxSize + 'px',
      overflow: textBoxOverflow,
    }"
    rows="1"
    @input="$emit('input', $event.target.value)"
    @click="$emit('click', $event)"
    @keyup="$emit('keyup', $event)"
    @keydown="$emit('keydown', $event)"
    @blur="$emit('blur', $event)"
  />
</template>
<script>
/**
 * A blank textarea with no surrounding box styling (see AutoExpandableTextareaInput if
 * you want a nice styled box around this) which will automatically expand given the
 * users input to the maximum number of rows specified by the maxRows prop.
 */
export default {
  name: 'AutoExpandableTextarea',
  props: {
    value: {
      required: true,
      type: String,
    },
    loading: {
      required: false,
      default: false,
      type: Boolean,
    },
    placeholder: {
      required: false,
      type: String,
      default: '',
    },
    maxRows: {
      required: false,
      type: Number,
      default: 4,
    },
    startingRows: {
      required: false,
      type: Number,
      default: 1,
    },
  },
  data() {
    return {
      numTextAreaLines: this.startingRows,
    }
  },
  computed: {
    textBoxSize() {
      return 22 * Math.min(this.numTextAreaLines, this.maxRows)
    },
    textBoxOverflow() {
      return this.numTextAreaLines > this.maxRows ? 'auto' : 'hidden'
    },
  },
  watch: {
    value() {
      this.resizeTextArea()
    },
  },
  mounted() {
    this.resizeTextArea()
  },
  methods: {
    resizeTextArea() {
      this.$nextTick(() => {
        const inputTextArea = this.$refs.inputTextArea
        this.numTextAreaLines = this.calculateHeight(inputTextArea)
      })
    },
    /**
     * Taken from https://stackoverflow.com/questions/1760629/how-to-get-number-of-rows-in-textarea-using-javascript/1761203#1761203
     *
     * The key reason we need this is to resize a fully expanded textarea to something
     * smaller as a user deletes newlines or text. Hence we need to actually manipulate
     * the dom and lower the height of the textarea until we find overflow is occurring
     * again to re-find the correct min height.
     */
    calculateContentHeight(ta, scanAmount) {
      const origHeight = ta.style.height
      let height = ta.offsetHeight
      const scrollHeight = ta.scrollHeight
      const overflow = ta.style.overflow
      /// only bother if the ta is bigger than content
      if (height >= scrollHeight) {
        /// check that our browser supports changing dimension
        /// calculations mid-way through a function call...
        ta.style.height = height + scanAmount + 'px'
        /// because the scrollbar can cause calculation problems
        ta.style.overflow = 'hidden'
        /// by checking that scrollHeight has updated
        if (scrollHeight < ta.scrollHeight) {
          /// now try and scan the ta's height downwards
          /// until scrollHeight becomes larger than height
          while (ta.offsetHeight >= ta.scrollHeight) {
            ta.style.height = (height -= scanAmount) + 'px'
          }
          /// be more specific to get the exact height
          while (ta.offsetHeight < ta.scrollHeight) {
            ta.style.height = height++ + 'px'
          }
          height--
          /// reset the ta back to it's original height
          ta.style.height = origHeight
          /// put the overflow back
          ta.style.overflow = overflow
          return height
        } else {
          // We weren't able to resize this element inside this function call to figure
          // out it's true height so we give up and just return the scrollHeight.
          return scrollHeight
        }
      } else {
        return scrollHeight
      }
    },
    calculateHeight(ta) {
      const style = window.getComputedStyle
        ? window.getComputedStyle(ta)
        : ta.currentStyle

      // This will get the line-height only if it is set in the css,
      // otherwise it's "normal"
      const taLineHeight = parseInt(style.lineHeight, 10)
      // Get the scroll height of the textarea
      const taHeight = this.calculateContentHeight(ta, taLineHeight)
      // calculate the number of lines
      return Math.ceil(taHeight / taLineHeight)
    },
  },
}
</script>
