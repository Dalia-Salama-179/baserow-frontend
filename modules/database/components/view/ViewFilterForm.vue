<template>
  <div>
    <div v-if="view.filters.length === 0">
      <div class="filters__none">
        <div class="filters__none-title">
          {{ $t('viewFilterContext.noFilterTitle') }}
        </div>
        <div class="filters__none-description">
          {{ $t('viewFilterContext.noFilterText') }}
        </div>
      </div>
    </div>
    <ViewFieldConditionsForm
      :filters="view.filters"
      :disable-filter="disableFilter"
      :filter-type="view.filter_type"
      :primary="primary"
      :fields="fields"
      :view="view"
      :read-only="readOnly"
      class="filters__items"
      @deleteFilter="deleteFilter($event)"
      @updateFilter="updateFilter($event)"
      @selectOperator="updateView(view, { filter_type: $event })"
    />
    <div v-if="!disableFilter" class="filters_footer">
      <a class="filters__add" @click.prevent="addFilter()">
        <i class="fas fa-plus"></i>
        {{ $t('viewFilterContext.addFilter') }}</a
      >
      <div v-if="view.filters.length > 0">
        <SwitchInput
          :value="view.filters_disabled"
          @input="updateView(view, { filters_disabled: $event })"
          >{{ $t('viewFilterContext.disableAllFilters') }}</SwitchInput
        >
      </div>
    </div>
  </div>
</template>

<script>
import { notifyIf } from '@baserow/modules/core/utils/error'
import ViewFieldConditionsForm from '@baserow/modules/database/components/view/ViewFieldConditionsForm'

export default {
  name: 'ViewFilterForm',
  components: {
    ViewFieldConditionsForm,
  },
  props: {
    primary: {
      type: Object,
      required: true,
    },
    fields: {
      type: Array,
      required: true,
    },
    view: {
      type: Object,
      required: true,
    },
    readOnly: {
      type: Boolean,
      required: true,
    },
    disableFilter: {
      type: Boolean,
      required: true,
    },
  },
  computed: {
    filterTypes() {
      return this.$registry.getAll('viewFilter')
    },
  },
  methods: {
    async addFilter(values) {
      try {
        await this.$store.dispatch('view/createFilter', {
          view: this.view,
          field: this.primary,
          values: {
            field: this.primary.id,
          },
          emitEvent: false,
          readOnly: this.readOnly,
        })
        this.$emit('changed')
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
    async deleteFilter(filter) {
      try {
        await this.$store.dispatch('view/deleteFilter', {
          view: this.view,
          filter,
          readOnly: this.readOnly,
        })
        this.$emit('changed')
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
    /**
     * Updates a filter with the given values. Some data manipulation will also be done
     * because some filter types are not compatible with certain field types.
     */
    async updateFilter({ filter, values }) {
      try {
        await this.$store.dispatch('view/updateFilter', {
          filter,
          values,
          readOnly: this.readOnly,
        })
        this.$emit('changed')
      } catch (error) {
        notifyIf(error, 'view')
      }
    },
    /**
     * Updates the view filter type. It will mark the view as loading because that
     * will also trigger the loading state of the second filter.
     */
    async updateView(view, values) {
      this.$store.dispatch('view/setItemLoading', { view, value: true })

      try {
        await this.$store.dispatch('view/update', {
          view,
          values,
          readOnly: this.readOnly,
        })
        this.$emit('changed')
      } catch (error) {
        notifyIf(error, 'view')
      }

      this.$store.dispatch('view/setItemLoading', { view, value: false })
    },
  },
}
</script>
