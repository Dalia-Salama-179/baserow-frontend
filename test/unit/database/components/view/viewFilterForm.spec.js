import { TestApp } from '@baserow/test/helpers/testApp'
import ViewFilterForm from '@baserow/modules/database/components/view/ViewFilterForm'

const primary = {
  id: 1,
  name: 'Name',
  order: 0,
  type: 'text',
  primary: true,
  text_default: '',
  _: {
    loading: false,
  },
}

const fields = [
  {
    id: 2,
    table_id: 196,
    name: 'Stars',
    order: 1,
    type: 'rating',
    primary: false,
    max_value: 5,
    color: 'dark-orange',
    style: 'star',
    _: {
      loading: false,
    },
  },
  {
    id: 3,
    table_id: 196,
    name: 'Flag',
    order: 2,
    type: 'rating',
    primary: false,
    max_value: 10,
    color: 'dark-red',
    style: 'heart',
    _: {
      loading: false,
    },
  },
]

const view = {
  id: 1,
  name: 'Grid',
  type: 'grid',
  filter_type: 'AND',
  filters: [
    {
      field: 1,
      type: 'equal',
      value: 'test',
      preload_values: {},
      _: { hover: false, loading: false },
      id: 10,
    },
    {
      field: 2,
      type: 'equal',
      value: 2,
      preload_values: {},
      _: { hover: false, loading: false },
      id: 11,
    },
  ],
  filters_disabled: false,
  _: {
    selected: true,
    loading: false,
  },
}

describe('ViewFilterForm component', () => {
  let testApp = null
  let mockServer = null

  beforeAll(() => {
    testApp = new TestApp()
    mockServer = testApp.mockServer
  })

  afterEach(() => {
    jest.useRealTimers()
    testApp.afterEach()
    mockServer.resetMockEndpoints()
  })

  const mountViewFilterForm = async (
    props = {
      primary: {},
      fields: [],
      view: { filters: [], _: {} },
      readOnly: false,
    },
    listeners = {}
  ) => {
    const wrapper = await testApp.mount(ViewFilterForm, {
      propsData: props,
      listeners,
    })
    return wrapper
  }

  test('Default view filter component', async () => {
    const wrapper = await mountViewFilterForm()
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Full view filter component', async () => {
    const wrapper = await mountViewFilterForm({
      primary,
      fields,
      view,
      readOnly: false,
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('Test rating filter', async (done) => {
    // We want to bypass some setTimeout
    jest.useFakeTimers()
    // Mock server filter update call
    mockServer.updateViewFilter(11, 5)

    // Add rating one filter
    const viewClone = JSON.parse(JSON.stringify(view))
    viewClone.filters = [
      {
        field: 2,
        type: 'equal',
        value: 2,
        preload_values: {},
        _: { hover: false, loading: false },
        id: 11,
      },
    ]

    const onChange = jest.fn(() => {
      // The test is about to finish
      expect(wrapper.emitted().changed).toBeTruthy()
      // The Five star option should be selected
      expect(wrapper.element).toMatchSnapshot()
      done()
    })

    // Mounting the component
    const wrapper = await mountViewFilterForm(
      {
        primary,
        fields,
        view: viewClone,
        readOnly: false,
      },
      { changed: onChange }
    )

    // Open type dropdown
    await wrapper.find('.filters__type .dropdown__selected').trigger('click')
    expect(wrapper.element).toMatchSnapshot()

    // Select five stars
    const option = wrapper.find(
      '.filters__value  .rating > .rating__star:nth-child(5)'
    )

    await option.trigger('click')
    // Wait some timers
    await jest.runAllTimers()

    // Test finishes only when onChange callback is called
    // Wait for mockServer to respond -> see onChange callback
  })
})
