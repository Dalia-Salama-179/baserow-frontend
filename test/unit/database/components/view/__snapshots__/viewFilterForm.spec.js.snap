// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewFilterForm component Default view filter component 1`] = `
<div>
  <div>
    <div
      class="filters__none"
    >
      <div
        class="filters__none-title"
      >
        
        viewFilterContext.noFilterTitle
      
      </div>
       
      <div
        class="filters__none-description"
      >
        
        viewFilterContext.noFilterText
      
      </div>
    </div>
  </div>
   
  <div
    class="filters__items"
  />
   
  <div
    class="filters_footer"
  >
    <a
      class="filters__add"
    >
      <i
        class="fas fa-plus"
      />
      
      viewFilterContext.addFilter
    </a>
     
    <!---->
  </div>
</div>
`;

exports[`ViewFilterForm component Full view filter component 1`] = `
<div>
  <!---->
   
  <div
    class="filters__items"
  >
    <div
      class="filters__item"
    >
      <a
        class="filters__remove"
      >
        <i
          class="fas fa-times"
        />
      </a>
       
      <div
        class="filters__operator"
      >
        <span>
          viewFilterContext.where
        </span>
         
        <!---->
         
        <!---->
      </div>
       
      <div
        class="filters__field"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        Name
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Name
      
                  </div>
                   
                  <!---->
                </a>
              </li>
               
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Stars
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Flag
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__type"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        viewFilter.is
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.is
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.isNot
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.contains
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.containsNot
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.lengthIsLowerThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.isEmpty
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.isNotEmpty
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__value"
      >
        <input
          class="input filters__value-input"
          type="text"
        />
      </div>
    </div>
    <div
      class="filters__item"
    >
      <a
        class="filters__remove"
      >
        <i
          class="fas fa-times"
        />
      </a>
       
      <div
        class="filters__operator"
      >
        <!---->
         
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        viewFilterContext.and
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <!---->
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilterContext.and
      
                  </div>
                   
                  <!---->
                </a>
              </li>
               
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilterContext.or
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
         
        <!---->
      </div>
       
      <div
        class="filters__field"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        Stars
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Name
      
                  </div>
                   
                  <!---->
                </a>
              </li>
               
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Stars
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Flag
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__type"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        viewFilter.is
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.is
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.isNot
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.higherThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.lowerThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__value"
      >
        <div
          class="filters__value-rating"
        >
          <div
            class="rating color--dark-orange editing"
          >
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star"
            />
            <i
              class="fas rating__star fa-star"
            />
            <i
              class="fas rating__star fa-star"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="filters_footer"
  >
    <a
      class="filters__add"
    >
      <i
        class="fas fa-plus"
      />
      
      viewFilterContext.addFilter
    </a>
     
    <div>
      <div
        class="switch switch--has-content"
      >
        viewFilterContext.disableAllFilters
      </div>
    </div>
  </div>
</div>
`;

exports[`ViewFilterForm component Test rating filter 1`] = `
<div>
  <!---->
   
  <div
    class="filters__items"
  >
    <div
      class="filters__item"
    >
      <a
        class="filters__remove"
      >
        <i
          class="fas fa-times"
        />
      </a>
       
      <div
        class="filters__operator"
      >
        <span>
          viewFilterContext.where
        </span>
         
        <!---->
         
        <!---->
      </div>
       
      <div
        class="filters__field"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        Stars
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Name
      
                  </div>
                   
                  <!---->
                </a>
              </li>
               
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Stars
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Flag
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__type"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex=""
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        viewFilter.is
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item active hover"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.is
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.isNot
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.higherThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.lowerThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__value"
      >
        <div
          class="filters__value-rating"
        >
          <div
            class="rating color--dark-orange editing"
          >
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star"
            />
            <i
              class="fas rating__star fa-star"
            />
            <i
              class="fas rating__star fa-star"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="filters_footer"
  >
    <a
      class="filters__add"
    >
      <i
        class="fas fa-plus"
      />
      
      viewFilterContext.addFilter
    </a>
     
    <div>
      <div
        class="switch switch--has-content"
      >
        viewFilterContext.disableAllFilters
      </div>
    </div>
  </div>
</div>
`;

exports[`ViewFilterForm component Test rating filter 2`] = `
<div>
  <!---->
   
  <div
    class="filters__items"
  >
    <div
      class="filters__item"
    >
      <a
        class="filters__remove"
      >
        <i
          class="fas fa-times"
        />
      </a>
       
      <div
        class="filters__operator"
      >
        <span>
          viewFilterContext.where
        </span>
         
        <!---->
         
        <!---->
      </div>
       
      <div
        class="filters__field"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        Stars
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Name
      
                  </div>
                   
                  <!---->
                </a>
              </li>
               
              <li
                class="select__item active"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Stars
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        Flag
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__type"
      >
        <div
          class="dropdown dropdown--floating dropdown--tiny"
          tabindex="0"
        >
          <a
            class="dropdown__selected"
          >
            <!---->
            
        viewFilter.is
       
            <i
              class="dropdown__toggle-icon fas fa-caret-down"
            />
          </a>
           
          <div
            class="dropdown__items hidden"
          >
            <div
              class="select__search"
            >
              <i
                class="select__search-icon fas fa-search"
              />
               
              <input
                class="select__search-input"
                placeholder="action.search"
                tabindex="0"
                type="text"
              />
            </div>
             
            <ul
              class="select__items"
              tabindex=""
            >
              <li
                class="select__item active hover"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.is
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.isNot
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.higherThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
              <li
                class="select__item"
              >
                <a
                  class="select__item-link"
                >
                  <div
                    class="select__item-name"
                  >
                    <!---->
                    
        viewFilter.lowerThan
      
                  </div>
                   
                  <!---->
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
       
      <div
        class="filters__value"
      >
        <div
          class="filters__value-rating"
        >
          <div
            class="rating color--dark-orange editing"
          >
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
            <i
              class="fas rating__star fa-star rating__star--selected"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    class="filters_footer"
  >
    <a
      class="filters__add"
    >
      <i
        class="fas fa-plus"
      />
      
      viewFilterContext.addFilter
    </a>
     
    <div>
      <div
        class="switch switch--has-content"
      >
        viewFilterContext.disableAllFilters
      </div>
    </div>
  </div>
</div>
`;
